### DISCUSSION TITLE
Use Google Typograms

### DISCUSSION THREAD

**Discussion by RduMarais** [marp-team/marp#467] [Ideas]
Hello, 

Thanks for all the work put in marp !
[Google Typograms](https://google.github.io/typograms/) is a lib to render diagrams in the browser.
It is very complete, powerful, and basically only needs 2 lines of code : 
```html
<script src="https://google.github.io/typograms/typograms.js"></script>
The diagram : 
<script type="text/typogram">
+----+
|    |---> My first diagram!
+----+
</script>
```

As of today, it works very well per se as HTML tags (with `--html`) in a mardown page rendered.
If other marp cli users would be interested, is it possible to include typogram as a language tag in marp ? such as 
`````
```typogram
+----+
|    |---> My first diagram!
+----+
```
`````
As far as i understand, it requires to build a dedicated engine as it is the case [in this issue](https://github.com/marp-team/marp-vscode/issues/116), but this needs install via npm wich is impossible for me + i think the strength of marp is its lightweight core.
I would be happy to contribute, but as of today i did not find any simple way to tweak this. 
Would you have any pointers in that direction ?

**Comment by yhatt** [comment]
You can write a small Marp plugin like this.

```javascript
// typograms-plugin.mjs
export const typogramsPlugin = (md) => {
  const { fence } = md.renderer.rules

  md.renderer.rules.fence = (tokens, i, options, env, self) => {
    const info = md.utils.unescapeAll(tokens[i].info).trim()

    if (info === 'typograms') {
      let output = `<script type="text/typogram">\n${tokens[i].content}</script>`

      if (!env.typogramsScriptLoaded) {
        output = `<script src="https://google.github.io/typograms/typograms.js"></script>${output}`
      }

      env.typogramsScriptLoaded = true

      return output
    }

    return fence.call(self, tokens, i, options, env, self)
  }
}
```

And you can use this plugin through [Marp CLI](https://github.com/marp-team/marp-cli#configuration-file):

```js
// marp.config.mjs
import { typogramsPlugin } from './typograms-plugin.mjs'

export default {
  engine: ({ marp }) => marp.use(typogramsPlugin),
}
```

```
marp --config ./marp.config.mjs typograms.md
```

> [!WARNING]
> In Marp's HTML export, typograms.js seems not to be able to recognize invisible diagrams that were put on initially inactive slide pages, because typograms.js is [using `Element.innerText`](https://github.com/google/typograms/blob/964dedc02e0c28bacbd27cade8857b34e15344d1/src/browser.js#L10C27-L10C36). (https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent#differences_from_innertext) If you concerned, recommend to send a feedback to [typograms issue tracker](https://github.com/google/typograms/issues).

**Comment by soaple** [comment]
Hi, I created markdown-it plugin for Google Typograms.
https://github.com/markslides/markdown-it-typograms

Also, you can Edit/Preview/Show your slide on MarkSlide, which is standalone PWA based on Marp.
https://www.markslides.ai/

<img width="1512" alt="image" src="https://github.com/marp-team/marp/assets/34570624/e12d78be-19c8-47de-a628-b40221f0c532">
